### ISSUE TITLE
Question about the solution for activity 5.01

### ISSUE THREAD

**Issue by xyangwu** [PacktPublishing/SQL-for-Data-Analytics-Third-Edition#4]
This issue concerns the calculation of deciles in the solution code.
https://github.com/PacktPublishing/SQL-for-Data-Analytics-Third-Edition/blob/78f7929e6e9519ddda16aae6737d65ddb649c881/Chapter05/Activity5.01/Activity5.01.sql#L39

The book presents the following question in activity 5.01:
> 4. Calculate which decile each date would fall into compared to other days based on their daily 30-day rolling sales amount.

Problem:
Line 68 of the solution code uses the window function `NTILE(10) OVER (ORDER BY moving_avg DESC) AS decile` to calculate the decile for each value group in the **moving_avg** column. However, the resulting **decile** column appears to be identical to what would be obtained using `NTILE(10) OVER () AS decile`, which omits the `ORDER BY` clause.

**Comment by junshan0** [comment]
Hi xyangwu, thank you for verifying the code!

I tried in my system but was not able to reproduce what you stated in the issue report. I suspect there might be two reasons:

1. It may just be coincidence that your intermediate data was already sorted (due to your processor and memory configuration), so even without the explicit ORDER BY, you are still getting the same result as if the ORDER BY is there.
2. It may also be that your version of PostgreSQL is capable of using the column before it for NTILE calculation. This is pure speculation as I do not know which version you are using. It is being a while since the book was published and PostgreSQL has released several new versions. I wouldn't be surprised if they added some new features.

That being said, I strongly recommend you define the ORDER BY for ALL windows functions, regardless of whether PostgreSQL can automatically determine the right option or not. The reasons (benefits) are:
1. Easy for future developers to understand. When you create your code, always keep in mind that after two or three years, there might be a new team member trying to read through to code and understand what is going on. Making your definitions explicit makes it easier for them to understand, and reduce the possibility of mistakes.
2. Easy for maintenance. Code change is unfortunately an inevitable part of life. Either the similarity in results is due to coincidence or PostgreSQL feature, when your code change, it is possible that the default result will change. Making your ORDER BY explicit makes it less likely that the result will change due to changes in outside environments.
3. Easy for migration. This is an extended case of #2. If your company moves to a different platform, you may not have the same result in the other platform with explicitly specifying the ORDER BY clause.

Overall, my point is, in your setup, the results could be the same, with or without the ORDER BY. However, for the long term benefit, it is recommended to have the ORDER BY clause instead of relying on certain PostgreSQL feature or coincidence. Hope this helps, and have a great day!

Jun

**Comment by xyangwu** [comment]
Thanks for the in-depth response! I'll definitely follow the coding principle you suggested.

Before raising this issue, I ran the following query in PostgreSQL versions 14 and 16, adding a new line `NTILE(10) OVER () AS decile_0` before line 68:

```SQL
SELECT
  sales_transaction_date,
  sales_amount,
  moving_avg,
  NTILE(10) OVER () AS decile_0,
  NTILE(10) OVER (ORDER BY moving_avg DESC) AS decile_desc
FROM 
  moving_avg m
WHERE 
  moving_avg IS NOT NULL
ORDER BY 
  decile_0;
```

Both **decile_0** and **decile_desc** columns produced the same results. Interestingly, removing the `ORDER BY` clause in the second window function (`NTILE(10) OVER (ORDER BY moving_avg DESC) AS decile_desc`) resulted in unordered deciles. It seems PostgreSQL might be using an implicit ordering in this case.

Initially, I was unsure about the behavior of aggregate functions in the context of window functions. I recalled the example of `COUNT(*)` used as a window function (Chapter 5, Page 194). There, `COUNT(*) OVER(ORDER BY customer_id)` creates a rolling window and applies the `COUNT` function within each window. This led me to explore the seeming confusing behavior of `NTILE`. However, unlike `COUNT`, NTILE itself doesn't have a rolling window mechanism. Now I know that it is a statistical function rather than aggregate function and it partitions the data into deciles based on the provided order (or implicit order).

Following your suggestion, I understand the importance of explicit `ORDER BY` clauses for predictable results in window functions. This clarifies how `NTILE` works and avoids reliance on implicit order.

Thank you again for your excellent explanation!
